File: C++/origin_version/segment_origin_version_02/test/resize_mask/resize.py
import numpy as np
import cv2

channels = 4
rows = 10
cols = 10
masks = np.arange(channels * rows * cols) + 100
masks = masks.reshape((4, 10, 10)).astype(np.float32)
print(masks.astype(np.int32))

dst_rows = 15
dst_cols = 15
masks = masks.transpose((1, 2, 0))
dst_masks = cv2.resize(masks, (dst_cols, dst_rows), interpolation=cv2.INTER_LINEAR)
if len(dst_masks.shape) == 2:
    dst_masks = dst_masks[:, :, None]

dst_masks = dst_masks.transpose((2, 0, 1))

print(dst_masks.astype(np.int32))

File: C++/origin_version/segment_origin_version_02/test/resize_mask/resize.cpp
#include <iostream>
#include <string>
#include <stdio.h>
#include <cuda_runtime.h>
#include "postprocess.h"

int main(){
    int num = 4;
    int rows = 10;
    int cols = 10;
    int size = num * rows * cols;
    float masks[size];
    for (int i = 0; i < size; i++){ masks[i] = i + 100; }
    for (int i = 0; i < size; i++){
        if (i % cols == 0) std::cout << std::endl;
        if (i % (rows * cols) == 0) std::cout << std::endl;
        std::cout << masks[i] << " ";
    }

    float* masksDevice = nullptr;
    cudaMalloc(&masksDevice, size * sizeof(float));
    cudaMemcpy(masksDevice, masks, size * sizeof(float), cudaMemcpyHostToDevice);

    int dst_rows = 15;
    int dst_cols = 15;
    int dst_size = num * dst_rows * dst_cols;
    float dst_masks[dst_size];

    float* dstMasksDevice = nullptr;
    cudaMalloc(&dstMasksDevice, dst_size * sizeof(float));

    resize(masksDevice, num, rows, cols, dstMasksDevice, dst_rows, dst_cols);

    cudaMemcpy(dst_masks, dstMasksDevice, dst_size * sizeof(float), cudaMemcpyDeviceToHost);

    for (int i = 0; i < dst_size; i++){
        if (i % dst_cols == 0) std::cout << std::endl;
        if (i % (dst_rows * dst_cols) == 0) std::cout << std::endl;
        std::cout << (int)dst_masks[i] << " ";
    }
    std::cout << std::endl;

    cudaFree(masksDevice);
    cudaFree(dstMasksDevice);

    return 0;
}
